Add 'cloneData' option to buildMemoryStorage

bhallionOhbibi opened the pull request:
See [issue 580](https://github.com/arthurfiorette/axios-cache-interceptor/issues/580) for more details

bhallionOhbibi commented:
Looks like there is some issue with the CI

arthurfiorette commented on src/storage/memory.ts:
```diff
@@ -31,19 +31,22 @@ declare const structuredClone: (<T>(value: T) => T) | undefined;
   cleanupInterval: number | false = false,
-  maxEntries: number | false = false
+  maxEntries: number | false = false,
+  storeClone = false
```
```suggestion
```

arthurfiorette commented on src/storage/memory.ts:
````diff
@@ -31,19 +31,22 @@ declare const structuredClone: (<T>(value: T) => T) | undefined;
  * ```
  *
  * @param {boolean} cloneData If the data returned by `find()` should be cloned to avoid
````
```suggestion
 * @param {boolean | 'double'} cloneData If the data returned by `find()` should be cloned to avoid
```

arthurfiorette replied:
Please change the documentation to reflect this changes too.

arthurfiorette commented on src/storage/memory.ts:
```diff
@@ -31,19 +31,22 @@ declare const structuredClone: (<T>(value: T) => T) | undefined;
  *
  * @param {boolean} cloneData If the data returned by `find()` should be cloned to avoid
- *   mutating the original data outside the `set()` method.
+ *   mutating the original data outside the `set()` method. Disabled is default
```
```suggestion
 *   mutating the original data outside the `set()` method. Use `'double'` to also clone `set()` return value. Disabled is default
```

arthurfiorette commented on src/storage/memory.ts:
```diff
@@ -67,6 +70,16 @@ export function buildMemoryStorage(
         }
       }
 
+      if (storeClone && value !== undefined) {
```
```suggestion
      if (cloneData === 'double' && value !== undefined) {
```

arthurfiorette commented on test/storage/memory.test.ts:
```diff
@@ -38,6 +38,30 @@ describe('tests memory storage', () => {
+  //
+  // https://github.com/arthurfiorette/axios-cache-interceptor/issues/580
+  it('not allow changes by value reference before', async () => {
+    const storage = buildMemoryStorage(false, false, false, true);
```
```suggestion
    const storage = buildMemoryStorage('double', false, false);
```

arthurfiorette commented on src/storage/memory.ts:
```diff
@@ -31,19 +31,22 @@ declare const structuredClone: (<T>(value: T) => T) | undefined;
+ *   future mutation from affecting cached data. Disabled is default
  */
 export function buildMemoryStorage(
   cloneData = false,
```
```suggestion
  cloneData: boolean | 'double' = false,
```

arthurfiorette commented on test/storage/memory.test.ts:
```diff
@@ -38,6 +38,30 @@ describe('tests memory storage', () => {
+    });
+
+    data.data = 'another data';
+
```
This check below checks if the `find()` clone is working. Please add expect statements with direct access to `storage.data`.

arthurfiorette commented:
The CI failed because of linting, please run `pnpm lint -- --fix` and `pnpm format` before push to lint your code.

arthurfiorette commented:
I added a review to update the `cloneData` property instead of creating a new one. Overall, suuper thanks for your time in this PR :)

bhallionOhbibi commented:
Thanks for your review. This is the first time writing typescript code. As yo ucan expect, my local configuration is not exactly mature right now :sweat_smile: 

I updated the test, but I'm not sure if this is enough, please let me know what you think.

About `storeClone`:
I removed it. It was an attempt to be a bit more explicit, as I find 'double' on `cloneData` less explicit. In the other hand, 'double' on `cloneData` limit amount of parameters and allow us to use the new option without adding `cleanupInterval` and `maxEntries` to false every time.
You decide what's best for this library and adding 'double' to `cloneData` seems like a fine typescript way of adding functionnality.

Another note about this whole clone thing. I really think that the `double` options should be the default value. I would argue that anyone trying this library for the first time might have a hard time understanding why their `axios.get` calls doesnt returns the same data every time. And suddenly having to check for `cached` property or having to provide additionnal configuration might block some users from using the library altogether.

arthurfiorette commented on test/storage/memory.test.ts:
```diff
@@ -56,8 +56,13 @@ describe('tests memory storage', () => {
-    expect(storage.data['key'].state).toBe('cached');
-    expect(storage.data['key'].data.data).toBe('data');
+    if (storage.data['key']) {
+      expect(storage.data['key'].state).toBe('cached');
```
```suggestion
      expect(storage.data['key']!.state).toBe('cached');
```

Use the `!` operator and remove these if checks.

arthurfiorette commented on src/storage/memory.ts:
```diff
@@ -67,6 +69,16 @@ export function buildMemoryStorage(
         }
       }
 
+      if (cloneData === 'double' && value !== undefined) {
```
```suggestion
      /* istanbul ignore if 'only available on super recent browsers' */
      if (cloneData === 'double' && value !== undefined) {
```

Add this comment to code coverage skip the structuredClone test

arthurfiorette commented on test/storage/memory.test.ts:
```diff
@@ -38,6 +38,39 @@ describe('tests memory storage', () => {
+  // a request to storage.get() should return unmodified value.
+  //
+  // https://github.com/arthurfiorette/axios-cache-interceptor/issues/580
+  it('not allow changes by value reference before', async () => {
```
```suggestion
  it('ensures set() also clones data when cloneData is double', async () => {
```

arthurfiorette commented:
> adding 'double' to cloneData seems like a fine typescript way

Yeah! I'm just not satisfied enough with the double name, any synonyms comes to your mind?

> I really think that the double options should be the default value. 

They may be, but i'd only make them default on the next major version as it will be a breaking change... Also, I'm not sure how the GC will behave on data being cloned everywhere.

bhallionOhbibi commented:
Instead of `double` I suggested `all`.

But it would be more explicit and also more configurable to use an enum.
Perhaps something like:
- `onSet`
- `onGet`

Where the user can specify one, both or none of the options (so an array of string maybe ?).
I think this aproach is the best, it wont break compatibility if you want to add clone to another method.
Let the user specify: `['onSet', 'onGet']`

I dont know how one would neatly implement this in typescript tho.

arthurfiorette commented on src/storage/memory.ts:
```diff
@@ -67,6 +69,17 @@ export function buildMemoryStorage(
       }
 
+      /* istanbul ignore if 'only available on super recent browsers' */
+      if (cloneData === 'double' && value !== undefined) {
```
```suggestion
      if (cloneData === 'double') {
```

value will never get undefined here

arthurfiorette commented on test/storage/memory.test.ts:
```diff
@@ -38,6 +38,35 @@ describe('tests memory storage', () => {
+    data.data = 'another data';
+
+    expect(storage.data['key']).not.toBeNull();
+    expect(storage.data['key']!.state).toBe('cached');
```
```suggestion
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    expect(storage.data['key']!.state).toBe('cached');
```

arthurfiorette commented on test/storage/memory.test.ts:
```diff
@@ -38,6 +38,35 @@ describe('tests memory storage', () => {
+
+    expect(storage.data['key']).not.toBeNull();
+    expect(storage.data['key']!.state).toBe('cached');
+    expect(storage.data['key']!.data).not.toBeNull();
```
```suggestion
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    expect(storage.data['key']!.data).not.toBeNull();
```

arthurfiorette commented on test/storage/memory.test.ts:
```diff
@@ -38,6 +38,35 @@ describe('tests memory storage', () => {
+    expect(storage.data['key']).not.toBeNull();
+    expect(storage.data['key']!.state).toBe('cached');
+    expect(storage.data['key']!.data).not.toBeNull();
+    expect(storage.data['key']!.data!.data).toBe('data');
```
```suggestion
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    expect(storage.data['key']!.data!.data).toBe('data');
```

arthurfiorette commented on test/storage/memory.test.ts:
```diff
@@ -38,6 +38,35 @@ describe('tests memory storage', () => {
+    const storage = buildMemoryStorage('double');
+
+    const data = { ...EMPTY_RESPONSE, data: 'data' };
+    await storage.set('key', {
```
```suggestion

    await storage.set('key', {
```

arthurfiorette reviewed (changes requested):
some linting

arthurfiorette commented:
Thanks!
